### DISCUSSION TITLE
Schema for representing dates (with time)

### DISCUSSION THREAD

**Discussion by vinay0000** [eose-tools-org/eose-api#3] [General]
The ``date`` object would be required to specify the mission epoch, orbit epoch, time-intervals (results of coverage calculations, ground-station/inter-satellite contact intervals), and in general, "events". 

The two required specifications for the ``date`` object are the: 
(1) format (e.g., ISO 8601, Julian Date) and (2) time-scales (e.g., UT1, UTC),

Below are time representation used by relevant software:

AstroPy: https://docs.astropy.org/en/stable/time/

Skyfield: https://rhodesmill.org/skyfield/time.html

GMAT: https://documentation.help/GMAT/SpacecraftEpoch.html

The associated ``date`` object could be restricted to ISO8601 (Gregorian Calendar) UTC and Julian date UT1. Example implementations:

```
"date": {
      "format": "ISO8601",
      "scale": "UTC",
      "value": "2002-204T15:56:23Z"
      },
```

```
 "date": {
      "format": "Julian_Date_UT1",
      "scale": "UT1",
      "value": 2459648.0
      },
```

@ptgrogan @ryanketzner @jsipps26 @dselva80 @bgorr

**Comment by ptgrogan** [comment]
Pydantic also has [datetime types](https://docs.pydantic.dev/2.0/usage/types/datetime/) that map to the Python datetime constructs. Dates/datetimes natively accept ISO 8601-like strings and Unix timestamps (milliseconds since 1 January 1970) and durations natively accept ISO 8601 duration strings and numbers (seconds).

Note: there are some small but important differences between ISO 8601 and various profiles of it including RFC 3339 used in various applications. In brief: serializing dates is more complicated than it seems.

Similar to the position discussion, it would generally be easier to choose one datetime format and require applications to do transformations to others on a as-needed basis. Choosing a "familiar" format such as UTC might be helpful because it is the most commonly implemented (i.e., no mapping implementation from UTC-UT1-Julian required). HOWEVER, UTC datetimes are much more difficult to interpret outside the bounds of 1972-present because of how leap seconds are defined.

**Reply by ptgrogan** [reply]
Also interestingly, there is a [2022 resolution](https://www.bipm.org/documents/20126/64811223/Resolutions-2022.pdf/281f3160-fc56-3e63-dbf7-77b76500990f?version=1.2&t=1668786143360) to stop defining UTC leap seconds after 2035!
